node/src/components/consensus/era_supervisor.rs

```diff
@@ -312,6 +325,26 @@ where
     pub(crate) fn active_eras(&self) -> &HashMap<EraId, Era<I>> {
         &self.active_eras
     }
+
+    /// To be called when we transition from the joiner to the validator reactor.
+    pub(crate) fn finished_joining(
+        &mut self,
+        now: Timestamp,
+    ) -> Vec<ProtocolOutcome<I, ClContext>> {
+        self.finished_joining = true;
+        let secret = Keypair::new(Rc::clone(&self.secret_signing_key), self.public_signing_key);
+        let public_key = self.public_signing_key;
+        self.active_eras
+            .get_mut(&self.current_era)
+            .map(|era| {
+                if era.start_time > now && era.validators().contains_key(&public_key) {
+                    era.consensus.activate_validator(public_key, secret, now)
```

Comment:
We haven't done it in the past b/c it can result in validator equivocating. If you had created some consensus messages in an era, then crashed/shut down/restarted and then tried rejoining then you will sync the linear chain and start producing consensus messages without any protocol state in that era. cc @afck @EdHastingsCasperLabs

Reply:
But it checks that `era.start_time > now`, so it can't equivocate, can it?
I was just going to suggest to use `era.start_time > self.node_start_time` here instead, so we are even more likely to start voting.

Reply:
Blind as a bat :see_no_evil: Shouldn't we at least verify if that era hasn't been deactivated already?

Reply:
How could it have been deactivated, if this is the first time an era gets activated in this node process?
I'll propose a slight modification of this in a PR later.

Reply:
No, wait, you're right:
We shouldn't activate the era if its switch block has already been finalized: Otherwise we'd never get the event for that, which is the only trigger to deactivate an era again…

Reply:
It checks if `era.start_time` is later than `now` trigger. Which means the switch block can't be finalized yet b/c era has not started yet :grimacing:

Reply:
Ah, right! And that's why we actually _shouldn't_ use `era.start_time > self.node_start_time`, I guess. :confused:

Reply:
We'll need to get back to this when we address [NDRS-647](https://casperlabs.atlassian.net/browse/NDRS-647). But in that case we'll probably have stored something about our own most recent message on disk.

Reply:
I think the only problem here is if we are a whale and the era that has already started can't progress without us - we _should_ start voting then, but we won't. Otherwise, this works fine: the validator will sync up to the current era, will follow the current era passively until the end and activate once the next era starts. Actually, the only case in which this branch activates is when we are a genesis validator started before genesis.

Reply:
> and the era that has already started can't progress without us - we should start voting then, 

I don't think we should, unless we persist the protocol state.

> the validator will sync up to the current era, will follow the current era passively until the end and activate once the next era starts. 

this should be already happening without changes in this PR. 

> Actually, the only case in which this branch activates is when we are a genesis validator started before genesis.

Same. It should already work this way without this PR :/

Reply:
> this should be already happening without changes in this PR.

One would think so ;) But the problem is that synchronization takes nonzero time. So, say, the node was started during era 71 and it started to sync then. Before it got up to era 71, consensus progressed to era 75. Once the node reached era 72, which started after the node was started, the check [here](https://github.com/CasperLabs/casper-node/blob/master/node/src/components/consensus/era_supervisor.rs#L254) started returning `false`, which caused the node to activate its `ActiveValidator` _while in the joiner state_ (because it was supposed to be a validator in era 72). This caused the consensus components to emit requests for blocks, which were dropped in the joiner and caused panics due to dropped responders.

Reply:
Oh yeah, I definitely agree about the new `FinishedJoining` logic :+1: